### PR TITLE
[SPARK-29283][SQL] Error message is hidden when query from JDBC, especially enabled adaptive execution

### PR DESCRIPTION
**Pull request by LantaoJin**
### What changes were proposed in this pull request?
When adaptive execution is enabled, the Spark users who connected from JDBC always get adaptive execution error whatever the under root cause is. It's very confused. We have to check the driver log to find out why.
```shell
0: jdbc:hive2://localhost:10000> SELECT * FROM testData join testData2 ON key = v;
SELECT * FROM testData join testData2 ON key = v;
Error: Error running query: org.apache.spark.SparkException: Adaptive execution failed due to stage materialization failures. (state=,code=0)
0: jdbc:hive2://localhost:10000> 
```

For example, a job queried from JDBC failed due to HDFS missing block. User still get the error message `Adaptive execution failed due to stage materialization failures`.

The easiest way to reproduce is changing the code of `AdaptiveSparkPlanExec`, to let it throws out  an exception when it faces `StageSuccess`.
```scala
  case class AdaptiveSparkPlanExec(
      events.drainTo(rem)
         (Seq(nextMsg) ++ rem.asScala).foreach {
           case StageSuccess(stage, res) =>
//            stage.resultOption = Some(res)
            val ex = new SparkException("Wrapper Exception",
              new IllegalArgumentException("Root cause is IllegalArgumentException for Test"))
            errors.append(
              new SparkException(s"Failed to materialize query stage: ${stage.treeString}", ex))
           case StageFailure(stage, ex) =>
             errors.append(
               new SparkException(s"Failed to materialize query stage: ${stage.treeString}", ex))
```

### Why are the changes needed?
To make the error message more user-friend and more useful for query from JDBC.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Manually test query:
```shell
0: jdbc:hive2://localhost:10000> CREATE TEMPORARY VIEW testData (key, value) AS SELECT explode(array(1, 2, 3, 4)), cast(substring(rand(), 3, 4) as string);
CREATE TEMPORARY VIEW testData (key, value) AS SELECT explode(array(1, 2, 3, 4)), cast(substring(rand(), 3, 4) as string);
+---------+--+
| Result  |
+---------+--+
+---------+--+
No rows selected (0.225 seconds)
0: jdbc:hive2://localhost:10000> CREATE TEMPORARY VIEW testData2 (k, v) AS SELECT explode(array(1, 1, 2, 2)), cast(substring(rand(), 3, 4) as int);
CREATE TEMPORARY VIEW testData2 (k, v) AS SELECT explode(array(1, 1, 2, 2)), cast(substring(rand(), 3, 4) as int);
+---------+--+
| Result  |
+---------+--+
+---------+--+
No rows selected (0.043 seconds)
```
Before:
```shell
0: jdbc:hive2://localhost:10000> SELECT * FROM testData join testData2 ON key = v;
SELECT * FROM testData join testData2 ON key = v;
Error: Error running query: org.apache.spark.SparkException: Adaptive execution failed due to stage materialization failures. (state=,code=0)
0: jdbc:hive2://localhost:10000> 
```
After:
```shell
0: jdbc:hive2://localhost:10000> SELECT * FROM testData join testData2 ON key = v;
SELECT * FROM testData join testData2 ON key = v;
Error: Error running query: java.lang.IllegalArgumentException: Root cause is IllegalArgumentException for Test (state=,code=0)
0: jdbc:hive2://localhost:10000> 
```

### CODE REVIEW

**Comment by LantaoJin** [comment]
Retest this please.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 1]
```diff
@@ -286,7 +286,8 @@ private[hive] class SparkExecuteStatementOperation(
           if (e.isInstanceOf[HiveSQLException]) {
             throw e.asInstanceOf[HiveSQLException]
           } else {
-            throw new HiveSQLException("Error running query: " + e.toString, e)
+            throw new HiveSQLException("Error running query: " +
+              SparkUtils.findFirstCause(e).toString, e)
```
Comment (wangyum) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala]: `SparkUtils.findFirstCause(e)` -> `org.apache.commons.lang3.exception.ExceptionUtils.getRootCause(e)`?

https://github.com/apache/commons-lang/blob/LANG_3_8_1/src/main/java/org/apache/commons/lang3/exception/ExceptionUtils.java#L167-L187

cc @juliuszsompolski @srowen

**Review by srowen — APPROVED** [review]

**Review by maropu — APPROVED** [review]

**Comment by juliuszsompolski** [comment]
For consistency, should we do that in all Spark*Operation?
I.e. replace the current
```
      case e: HiveSQLException =>
        setState(OperationState.ERROR)
        HiveThriftServer2.listener.onStatementError(
          statementId, e.getMessage, SparkUtils.exceptionString(e))
        throw e
```
with 
```
  case e: Throwable =>
    logError(s"Error executing operation with $statementId, currentState $currentState, ", e)
    setState(OperationState.ERROR)
    HiveThriftServer2.listener.onStatementError(
      statementId, e.getMessage, SparkUtils.exceptionString(e))
    if (e.isInstanceOf[HiveSQLException]) {
      throw e.asInstanceOf[HiveSQLException]
    } else {
      val root = ExceptionUtils.getRootCause(e)
      throw new HiveSQLException("Error running query: " +
        (if (root == null) e.toString else root.toString), e)
    }
```
in all of them?

**Comment by LantaoJin** [comment]
@juliuszsompolski fixed.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetCatalogsOperation.scala [hunk 2]
```diff
@@ -68,11 +69,18 @@ private[hive] class SparkGetCatalogsOperation(
       }
       setState(OperationState.FINISHED)
     } catch {
-      case e: HiveSQLException =>
+      case e: Throwable =>
```
Comment (maropu) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetCatalogsOperation.scala]: NonFatal?

Comment (juliuszsompolski) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetCatalogsOperation.scala]: Hm. I guess we may want to catch a Throwable.
E.g. InterruptedExpression is not catched by NonFatal, and we want to inform the HiveThriftServer2.listener about the error after an interrupt - this definitely can happen in SparkExecuteStatementOperation that is async and can be cancelled. After a ThreadDeath of OutOfMemoryError I think we also want to inform the HiveThriftServer2.listener to not get the query hanging in the UI, as I think the server would continue to go on (I think it won't bring the whole JVM down?).

Comment (maropu) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetCatalogsOperation.scala]: If so, we should list up `InterruptedExpression` here, too? IIUC the reason why we mainly use `NonFatal` in this case is not to catch `NonLocalReturnControl`.  But, yea, this is not my area, so I think @wangyum could suggest more about this.

Comment (wangyum) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetCatalogsOperation.scala]: +1 for `Throwable`.

Extractor of non-fatal Throwables. Will not match fatal errors like `VirtualMachineError`.
(for example, `OutOfMemoryError` and `StackOverflowError`, subclasses of `VirtualMachineError`), `ThreadDeath`, `LinkageError`, `InterruptedException`, `ControlThrowable`.

https://github.com/scala/scala/blob/v2.12.10/src/library/scala/util/control/NonFatal.scala#L17-L19

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala [hunk 3]
```diff
@@ -286,7 +287,9 @@ private[hive] class SparkExecuteStatementOperation(
           if (e.isInstanceOf[HiveSQLException]) {
             throw e.asInstanceOf[HiveSQLException]
           } else {
-            throw new HiveSQLException("Error running query: " + e.toString, e)
+            val root = ExceptionUtils.getRootCause(e)
```
Comment (wangyum) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala]: Could we change it to?
```scala
setState(OperationState.ERROR)
e match {
  case hiveException: HiveSQLException =>
    logError(s"Error executing query with $statementId, currentState $currentState, ", e)
    HiveThriftServer2.listener.onStatementError(
      statementId, hiveException.getMessage, SparkUtils.exceptionString(hiveException))
    throw hiveException
  case _ =>
    val rootCause = Option(ExceptionUtils.getRootCause(e)).getOrElse(e)
    logError(
      s"Error executing query with $statementId, currentState $currentState, ", rootCause)
    HiveThriftServer2.listener.onStatementError(
      statementId, rootCause.getMessage, SparkUtils.exceptionString(rootCause))
    throw new HiveSQLException("Error running query: " + rootCause.toString, rootCause)
}
```

Comment (LantaoJin) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala]: > ```scala
> val rootCause = Option(ExceptionUtils.getRootCause(e)).getOrElse(e)
> ```

Return null only if the input `e` is null. Do we still add this option?

Comment (LantaoJin) [reply sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala]: Besides the null checker, I've changed the code to above style. @wangyum

**Comment by LantaoJin** [comment]
retest this please.

**Comment by LantaoJin** [comment]
Retest this please.

**Comment by LantaoJin** [comment]
Retest this please.

**Comment by LantaoJin** [comment]
Retest this please.

**Comment by LantaoJin** [comment]
Retest this please.

**Comment by LantaoJin** [comment]
The UT could passed after #26090 merged.

**Comment by wangyum** [comment]
retest this please

**Comment by LantaoJin** [comment]
Retest this please.

**Review by wangyum — APPROVED** [review]
LGTM cc @juliuszsompolski @srowen @maropu

**Review by srowen — COMMENTED** [review]
I guess it's hard to refactor this error handling vs copying it? seems OK.

**Comment by juliuszsompolski** [comment]
LGTM.
@srowen yeah, some common logic could be refactored out into a mixin trait (because direct inheritence goes from different Hive Operation implementations), but I think it's a bigger refactor change than the scope of this PR.

**Comment by wangyum** [comment]
Merged to master and branch-3.0-preview.

**Pull request closed without merging**